_includes/manuals/1.17/7.2_debugging.md

```diff
@@ -30,6 +40,138 @@ Plugins may add their own logs too, see /etc/foreman/plugins/ for their possible
 
 Also see [Configuration Options]({{ site.baseurl }}manuals/{{ page.version }}/index.html#3.5.2ConfigurationOptions) for more information.
 
+#### Structured logging
+
+Starting from Foreman 1.18, logging stack can be configured to log into system journal:
+
+    :logging:
+      :level: info
+      :type: journald
+
+On Red Hat compatible systems, journald is running in transient mode by default and forwards all logs to syslog which means structured information is dropped after some time (memory buffer only holds few hours back). To see structured fields:
+
+    journalctl -fo verbose
+
+Here is the list of most important structured fields which can help with debugging:
+
+<table class="table table-bordered table-condensed">
+  <tr>
+    <th>Journald name</th>
+    <th>JSON name</th>
+    <th>Description</th>
+  </tr>
+  <tr>
+    <td>USER_LOGIN</td>
+    <td>["mdc"]["user_login"]</td>
+    <td>User login name</td>
+  </tr>
+  <tr>
+    <td>ORG_ID</td>
+    <td>["mdc"]["org_id"]</td>
+    <td>Organization database ID</td>
+  </tr>
+  <tr>
+    <td>LOC_ID</td>
+    <td>["mdc"]["loc_id"]</td>
+    <td>Location database ID</td>
+  </tr>
+  <tr>
+    <td>REMOTE_IP</td>
+    <td>["mdc"]["remote_ip"]</td>
+    <td>Remote IP address of a client</td>
+  </tr>
+  <tr>
+    <td>REQUEST</td>
+    <td>["mdc"]["request"]</td>
+    <td>Request ID generated by ActionDispatch</td>
+  </tr>
+  <tr>
+    <td>SESSION</td>
+    <td>["mdc"]["session"]</td>
+    <td>Random ID generated per session or request for session-less request</td>
+  </tr>
+  <tr>
+    <td>EXCEPTION_MESSAGE</td>
+    <td>["ndc"]["exception_message"]</td>
+    <td>Exception message when error is logged</td>
+  </tr>
+  <tr>
+    <td>EXCEPTION_CLASS</td>
+    <td>["ndc"]["exception_class"]</td>
+    <td>Exception Ruby class when error is logged</td>
+  </tr>
+  <tr>
+    <td>EXCEPTION_BACKTRACE</td>
+    <td>["ndc"]["exception_backtrace"]</td>
+    <td>Exception backtrace as a multiline string when error is logged</td>
+  </tr>
+  <tr>
+    <td>TEMPLATE_NAME</td>
+    <td>["ndc"]["template_name"]</td>
+    <td>Template name (blob logger)</td>
+  </tr>
+  <tr>
+    <td>TEMPLATE_CONTEXT</td>
+    <td>["ndc"]["template_context"]</td>
+    <td>Context of template renderer - Ruby class name (blob logger)</td>
+  </tr>
+  <tr>
+    <td>TEMPLATE_DIGEST</td>
+    <td>["ndc"]["template_digest"]</td>
+    <td>Digest (SHA256) of rendered template contents (blob logger)</td>
+  </tr>
+  <tr>
+    <td>TEMPLATE_HOST_NAME</td>
+    <td>["ndc"]["template_host_name"]</td>
+    <td>Host name for a rendered template if present (blob logger)</td>
+  </tr>
+  <tr>
+    <td>TEMPLATE_HOST_ID</td>
+    <td>["ndc"]["template_host_id"]</td>
+    <td>Host database ID for a rendered template if present (blob logger)</td>
+  </tr>
```

Comment:
not sure how to best formulate it, and how exactly it translates into journal name, but audit logger adds changed attributes under ndc, e.g. ["ndc"]["name"] would give you 'Default Organization' when new org is created, or it would give you back an array of ['Default Organization', 'New Name'] in case the name has changed.

```
<tr>
    <td>?</td>
    <td>["ndc"][$attribute]</td>
    <td>For every audited change, all attribute changes are logged. The key under ndc is changed attribute name, the value is either the value itself in case of resource creation, or an array of two values in case of resource update. The first item in this array is the old value, the second is the new value in this case. (audit logger)</td>
</tr>
```

Reply:
Hmmm, I don't know if this is actually ideal for ELK users for easy searching. Also, where do I track resource name (user, subnet, ...)? Wouldn't be better to be more explicit:

* AUDIT_TYPE=Create|Update|Delete
* AUDIT_RESOURCE=User
* AUDIT_ATTRIBUTE=login
* AUDIT_OLD=lzap
* AUDIT_NEW=lzap80

Is this feasible? I can do the PR if there is not a technical limitation to do that.

Reply:
It is feasible and shouldn't be too hard to do. But note that we're generating one log atm for the whole hash of changes. With your suggestion, it seems as separate log per attribute, so if I change name and desc, that generates two logs. If that's not concerning, I'm happy to review. See the introducing PR you co-reviewed for directions and ping me if you need any help. I've heard 1.18 branching is not gonna happen until tomorrow morning :-)